simpleast-core/src/main/java/com/discord/simpleast/core/parser/ParseSpec.kt

```diff
@@ -17,18 +17,21 @@ import com.discord.simpleast.core.node.Node
 class ParseSpec<R, T : Node<R>> {
   val root: T?
   val isTerminal: Boolean
+  val state: Map<String, Any>
```

Comment:
This was kind of what I was thinking, However I would try and avoid using maps since those are costly to build and state is probably going to be simple. As most rules probably won't be looking at state, could we use have state be an "Any" type? If the Rules need it, then they can do a type check and a cast, which they already need to do for the value in the map. Going further I would like to make it generic if possible. I was even thinking maybe we could use bit flags, but that's probably too restrictive...

My hunch is that we should do something like `RenderContext` (the `R` generic type here) and use interfact + composition to define acceptable types